### PR TITLE
docs: replace extend any with unknown for no-unnecessary-type-constra…

### LINKED ISSUE #6973: Docs: no-unnecessary-type-constraint should mention extending unknown
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

@MichaelDeBoey raises an excellent point in https://github.com/remix-run/remix/pull/6248#discussion_r1181130274:

> The docs are talking about `any` only, but this had `unknown` Does the same apply? If so, I guess it would be a good idea to update the docs?

Agreed - we should update the docs. They should mention when `extends unknown` is unnecessary, and include at least one or two examples in the code blocks. Accepting PRs!

### Affected URL(s)

https://typescript-eslint.io/rules/no-unnecessary-type-constraint/

### PR DESCRIPTION
**Pull request by takanome-dev**
## PR Checklist

- [x] Addresses an existing open issue: fixes #6973 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

This pr replaces the default constraint for `no-unnecessary-type-constraint` from `any` to `unknown` in the docs.

Note: I'm not really sure if I have to change all the any to unknown in the example block. I tried in the test file and they all failed :sweat: I have added two more examples using unknown (only in the incorrect section).

Please do let me know if there are additional steps I need to take :pray:

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/no-unnecessary-type-constraint.md [hunk 1]
```diff
@@ -7,8 +7,8 @@ description: 'Disallow unnecessary constraints on generic types.'
 > See **https://typescript-eslint.io/rules/no-unnecessary-type-constraint** for documentation.
 
 Generic type parameters (`<T>`) in TypeScript may be "constrained" with an [`extends` keyword](https://www.typescriptlang.org/docs/handbook/generics.html#generic-constraints).
-When no `extends` is provided, type parameters default a constraint to `any`.
-It is therefore redundant to `extend` from `any`.
+When no `extends` is provided, type parameters default a constraint to `unknown`.
+It is therefore redundant to `extend` from `unknown`.
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/no-unnecessary-type-constraint.md:11]: [Phrasing] Ah, the rule reports both on `any` and `unknown` as of TypeScript 3.9:

https://github.com/typescript-eslint/typescript-eslint/blob/09c04fb0d5a5dee0dc266abb2e21aa9a1c489712/packages/eslint-plugin/src/rules/no-unnecessary-type-constraint.ts#L56-L61

How about ...

```suggestion
It is therefore redundant to `extend` from `any` or `unknown`.
```

File: packages/eslint-plugin/docs/rules/no-unnecessary-type-constraint.md [hunk 2]
```diff
@@ -19,8 +19,12 @@ It is therefore redundant to `extend` from `any`.
 ```ts
 interface FooAny<T extends any> {}
 
+interface FooUnknown<T extends unknown> {}
+
 type BarAny<T extends any> = {};
 
+type BarUnknown<T extends unknown> = {};
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/no-unnecessary-type-constraint.md:26]: [Praise] Just confirming, yes, this looks reasonable to me 😄. We want both the existing `extends any`s and `extends unknown`s.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Sorry this took so long @takanome-dev, I'd meant to enthusiastically respond but it slipped by. Thanks for sending!

Mostly looks great - just one suggestion for the docs to make the phrasing more clear. Thoughts? 

Heidi Gardner dressed in fashion in SNL's Weekend Update enthusiastically saying 'I love it!'

File: packages/eslint-plugin/docs/rules/no-unnecessary-type-constraint.md (hunk 1 above)
Comment (takanome-dev) [reply packages/eslint-plugin/docs/rules/no-unnecessary-type-constraint.md:11]: Yeah, make sense :+1:

File: packages/eslint-plugin/docs/rules/no-unnecessary-type-constraint.md (hunk 2 above)
Comment (takanome-dev) [reply packages/eslint-plugin/docs/rules/no-unnecessary-type-constraint.md:26]: :smiley:

**Comment by takanome-dev** [comment]
> Mostly looks great - just one suggestion for the docs to make the phrasing more clear. Thoughts?
> 

Sounds good :+1: Thanks for taking the time reviewing this :slightly_smiling_face:

**Review by JoshuaKGoldberg — APPROVED** [review]
Lady in red dress emphatically gesturing and saying 'flawless'

**Pull request merged**